### ISSUE TITLE
PubSub message.ack() has no effect after calling subscription.removeListener

### ISSUE THREAD

**Issue by ChrisBartonLC** [GoogleCloudPlatform/google-cloud-node#2568]
Calling 'subscription.removeListener('message', on_message)' causes 'message.ack()' to have no effect.

**This doesn't work**
subscription.on('message', on_message)
...
subscription.removeListener('message', on_message)
...
messages.forEach(message => message.ack())

**This works**
subscription.on('message', on_message)
...
messages.forEach(message => message.ack())
...
subscription.removeListener('message', on_message)

**Comment by callmehiphop** [comment]
So we open a bi-directional stream when event listeners are added to a subscription and when the listeners are removed we close the stream. In the scenario above we end up batching together all the `message.ack()` calls to send through a single rpc call. 

Are you running this in a cloud function? Because we are using a 1 second timeout to batch together all the messages, in which case I would suggest removing the listener after acking.

**Comment by chris-barton-localcover** [comment]
Yep I am using a CloudFunction. So is it expected behaviour that you cannot call message.ack() after the 'message' listener has been removed? It all works if I just comment out the call to 'removeListener' because I have a delay of a few seconds when closing the subscription.

**Comment by callmehiphop** [comment]
No, the expectation is that you can call `message.ack()` after removing the message listeners, but it will take a little longer (about 1 second or so) to actually ack the message. I'm unable to reproduce this bug locally, so if you wouldn't mind putting together some code to replicate the bug that would be awesome and very much appreciated.

**Comment by lukesneeringer** [comment]
@callmehiphop Would it be reasonable to have the stream closure (after the final listener is removed) be gated on a flush of the acks?

**Comment by ChrisBartonLC** [comment]
OK problem reproduced

Finished listening for messages after 15000 milliseconds
Acknowledging first message from a total of 5 hi3
Removing subscription listener...
Acknowledging message after subscription listener removed hi3
Acknowledging message after subscription listener removed hi3
Acknowledging message after subscription listener removed hi3
Acknowledging message after subscription listener removed hi3
Acknowledging message after subscription listener removed hi3
Watiing for subscription to close...
Finished waiting for subscription to close after 3000 milliseconds
Macintosh:fn fo$ node -e 'require("./debug-pubsub").debug()'
Listening for messages...
Retrieved message hi
Retrieved message hi4
Retrieved message hi5
Retrieved message hi2
Finished listening for messages after 15000 milliseconds
Acknowledging first message from a total of 4 hi
Removing subscription listener...
Acknowledging message after subscription listener removed hi
Acknowledging message after subscription listener removed hi
Acknowledging message after subscription listener removed hi
Acknowledging message after subscription listener removed hi
Watiing for subscription to close...
Finished waiting for subscription to close after 3000 milliseconds

**Comment by ChrisBartonLC** [comment]
'use strict'

/**
 * ----------------------------------------------------------------------
 * Usage:
 * 1) gcloud beta pubsub topics create tmp
 * 2) gcloud beta pubsub subscriptions create tmp --topic=tmp --ack-deadline=10
 * 3) node -e 'require("./debug-pubsub").debug()'
 *
 * Add a message
 * gcloud beta pubsub topics publish tmp hi
 *
 * Read messages from the topic
 * gcloud beta pubsub subscriptions create tmp --topic=tmp_tmp --ack-deadline=10
 * gcloud --format=json beta pubsub subscriptions pull tmp_tmp --max-messages=10
 * -> echo "aGk=" | base64 --decode
 *
 * ----------------------------------------------------------------------
 */

module.exports.debug = function () {
  const pubsub = require('@google-cloud/pubsub')({ projectId: 'localcover-55' })

  const topic = pubsub.topic('tmp')
  const subscription = topic.subscription('tmp', {flowControl: { maxMessages: 5 }})

  console.log('Listening for messages...')
  subscription.on('error', function (err) {
    console.error('Error processing messages', err)
    subscription.removeListener('message', on_message)
  })
  subscription.on('message', on_message)

  function subscription_close(subscription) {
    const timeout = 3000

    console.log('Watiing for subscription to close...')
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        console.log(`Finished waiting for subscription to close after ${timeout} milliseconds`)
        subscription.close().then(resolve, reject)
      }, timeout)
    })
  }

  let messages = []

  function on_message(message) {
    let payload = Buffer.from(message.data, 'base64').toString()
    console.log('Retrieved message', payload)

    messages.push(message)
  }

  const timeout = 15000

  return new Promise(
    function (resolve, reject) {

      setTimeout(function () {
        console.log(`Finished listening for messages after ${timeout} milliseconds`)

        if (messages.length > 0) {
          console.log(`Acknowledging first message from a total of ${messages.length}`, Buffer.from(messages[0].data, 'base64').toString())
          messages[0].ack()
        }

        console.log('Removing subscription listener...')
        subscription.removeListener('message', on_message)

        for (let m of messages) {
          console.log('Acknowledging message after subscription listener removed', Buffer.from(messages[0].data, 'base64').toString())
          m.ack()
        }

        subscription_close(subscription)
          .then(() => resolve())
          .catch(err => reject(err))

      }, timeout)

    })
}

**Comment by ChrisBartonLC** [comment]
If I also add a pause prior to closing the subscription in addition to pausing after the subscription is closing the behaviour is the same

**Comment by callmehiphop** [comment]
While this issue is unresolved, I'm closing it because the core issue is a duplicate of #2558.
